Improve location breadcrumbs

esoergel opened the pull request:
Make "Locations" take you to the current location's position in the tree, and add a breadcrumb for the parent location's edit page.

![](https://i.imgur.com/a43810e.png)
@sravfeyn

esoergel committed:
Combine urlecode and reverse

dimagimon commented on corehq/util/__init__.py:
```diff
@@ -1,5 +1,7 @@
 import re
 
+from .view_utils import reverse
```
F401 'reverse' imported but unused

esoergel replied:
au contraire, @dimagimon, au contraire

benrudolph replied:
haha how is this being used?

czue replied:
@benrudolph I think he's just using the convention of putting the public API of a module into the `__init__.py` file.

esoergel replied:
Yup.
```python
# without internal linking
from corehq.utils.view_utils import reverse
from corehq.utils.dates import unix_time

# if corehq.utils.__init__ imports those two, you can just write
from corehq.utils import reverse, unix_time
```
It lets you present a friendly public API while still having a well organized module.

esoergel committed:
Improve location breadcrumbs

Make "Locations" take you to the current location's position in the
tree, and add a breadcrumb for the parent location's edit page.

czue commented on corehq/apps/locations/views.py:
```diff
@@ -349,6 +349,26 @@ class EditLocationView(NewLocationView):
+                    args=[self.domain, self.location.parent_id],
+                )
+            })
+        return breadcrumbs
```
does this intentionally only go up two levels? seems like it could be nice to show the whole thing (which might be easiest to just do recursively)

esoergel replied:
I thought about it, decided this was the simplest, but if there's interest, I'd be up for adding in the rest of the parents.
As-is this just shows a link to all locations and a link to the immediate parent.

esoergel replied:
https://github.com/dimagi/commcare-hq/pull/6352

czue commented:
:+1: from me. I think there are several places we don't use `absolute_reverse` where we probably should (e.g. email reports)

sravfeyn commented on corehq/util/view_utils.py:
```diff
@@ -107,5 +107,18 @@ def get_request():
     return global_request.get_request()
 
 
+def reverse(viewname, params=None, absolute=False, **kwargs):
```
Fine to name this same as django's own reverse?

esoergel replied:
Because this is backwards-compatible, I think it's fine.  If you need to pass parameters, you can just swap out the import statement without worrying about breaking the other usages.
